src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```diff
@@ -191,58 +186,104 @@ public override void Return(T[] array, bool clearArray = false)
             }
 
+            // Trim each of the TLS buckets. Note that threads may be modifying their TLS slots concurrently with
+            // this trimming happening. We do not force synchronization with those operations, so we accept the fact
+            // that we may end up firing a trimming event even if an array wasn't trimmed, and potentially
+            // trim an array we didn't need to.  Both of these should be rare occurrences.
+
             if (pressure == Utilities.MemoryPressure.High)
             {
-                // Under high pressure, release all thread locals
-                if (log.IsEnabled())
+                // Under high pressure, release all thread locals.
+                if (!log.IsEnabled())
                 {
-                    foreach (KeyValuePair<T[]?[], object?> tlsBuckets in _allTlsBuckets)
+                    foreach (KeyValuePair<ThreadLocalArray[], object?> tlsBuckets in _allTlsBuckets)
                     {
-                        T[]?[] buckets = tlsBuckets.Key;
+                        Array.Clear(tlsBuckets.Key);
+                    }
+                }
+                else
+                {
+                    foreach (KeyValuePair<ThreadLocalArray[], object?> tlsBuckets in _allTlsBuckets)
+                    {
+                        ThreadLocalArray[] buckets = tlsBuckets.Key;
                         for (int i = 0; i < buckets.Length; i++)
                         {
-                            T[]? buffer = Interlocked.Exchange(ref buckets[i], null);
-                            if (buffer is not null)
+                            if (Interlocked.Exchange(ref buckets[i].Array, null) is T[] buffer)
                             {
-                                // As we don't want to take a perf hit in the rent path it
-                                // is possible that a buffer could be rented as we "free" it.
                                 log.BufferTrimmed(buffer.GetHashCode(), buffer.Length, Id);
                             }
                         }
                     }
                 }
-                else
+            }
+            else
+            {
+                // Otherwise, release thread locals based on how long we've observed them to be stored. This time is
+                // approximate, with the time set not when the array is stored but when we see it during a Trim, so it
+                // takes at least two Trim calls (and thus two gen2 GCs) to drop an array, unless we're in high memory
+                // pressure. These values have been set arbitrarily; we could tune them in the future.
+                uint millisecondsThreshold = pressure switch
+                {
+                    Utilities.MemoryPressure.Medium => 15_000,
+                    _ => 30_000,
+                };
+
+                foreach (KeyValuePair<ThreadLocalArray[], object?> tlsBuckets in _allTlsBuckets)
```

Comment:
More a question than anything else:

The `Trim()` code has grown relatively complex. Is it fine to run all this on the finalizer thread? Wouldn't it be better to delegate the actual work to ThreadPool? Or is it negligible considering the whole queue?

Reply:
It's not adding a lot of work.  It'd be an ok thing to explore separately, but queueing it to the thread pool would also enable concurrent invocations of Trim, and the implementation would need to be audited to make sure that worked correctly.  It would also mean that trimming could be delayed not just by lots of objects to be finalized but also by lots of queued work to the pool, which could have harder to understand emergent behavior.  And this only fired because we're doing a gen2 GC, which means we're already doing a lot of work as part of that.